### DISCUSSION TITLE
Node.js logflare logs from serverless function without vercel integration or pino.

### DISCUSSION THREAD

**Discussion by dbauszus-glx** [vercel/vercel#5204] [Show and tell]
We want to log from inside our serverless function execution but without going through the hoops of setting up pino and pino.logflare with streams to the vercellogflare module.

A quick and dirty logger module we came up with looks like this:

```javascript
const fetch = require('node-fetch')

module.exports = msg => {
  
  fetch(`https://api.logflare.app/logs/json?api_key=${process.env.KEY_LOGFLARE}&source=${process.env.SOURCE_TOKEN}`,
  {
    method: 'post',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      msg: msg
    })
  }).then(resp=>{
    !resp.ok && console.log(msg)
  }).catch(err=>{
    console.error(err)
  })

}
```

A message send to the logger module will be posted to our logflare drain given a valid logflare api key and source token.

If unsuccessful the message will be logged to the console.

This might be helpful to some of you working with the node.js runtime for serverless functions.

**Comment by paulogdm** [comment]
The proper solution is to use https://vercel.com/blog/log-drains. This will allow you to effectively query your logs with ease.

**Reply by psn87** [reply]
+

**Comment by dbauszus-glx** [comment]
@paulogdm Getting a logflare log-drain setup for the build step and edge network was fairly straightforward.

Runtime however proved more challenging and requires to setup pino logger and [pino-logflare](https://github.com/Logflare/pino-logflare). The entire documentation available documentation for this is [here](https://github.com/Logflare/pino-logflare/blob/master/docs/VERCEL).

I would give it another try if there is a working example to plug into a custom node.js function.

But I was not able to get this working in a reasonable amount of time given that we do not have a lot experience with pino, or logflare. Which is why we decided after 2 hours to put the above solution into effect which took us less than 10 minutes to write and implement and does not require additional dependencies.

**Reply by paulogdm** [reply]
You shouldn't need "pino" to receive logs from your functions. All setups that I have done to date are not using any libraries. The platform will send you 'raw logs' and you can process it as you wish with Logflare queries & filters.

**Reply by dbauszus-glx** [reply]
We had a log drain set up but did not get anything apart from the build logs and the requests. Console.log messages we could see in the Vercel functions interface but not transmitted to the log flare source.

**Reply by paulogdm** [reply]
Did you look for "message"? Check the following link: https://vercel.com/docs/api#integrations/log-drains/ndjson-drains

**Reply by dbauszus-glx** [reply]
@paulogdm you are right. The message is there. I just gave this another try. Putting the req.url in a console log I get this in my logflare log drain.

```
 {
  "deploymentId": "dpl_8WKEK4riAVjo2FxT6n3JXNjRSAHd",
  "host": "xyz-mapbox-jvr7tqifm.vercel.app",
  "id": "1600939544169781952179534331",
  "message": "START RequestId: 37854802-1647-4a7a-bc0b-0e15849e4831 Version: $LATEST\n2020-09-24T09:25:44.234Z\t37854802-1647-4a7a-bc0b-0e15849e4831\tINFO\t/api/workspace/get/layer?locale=Mapbox&key=layer&layer=Mapbox%20Static&method=get\nEND RequestId: 37854802-1647-4a7a-bc0b-0e15849e4831\nREPORT RequestId: 37854802-1647-4a7a-bc0b-0e15849e4831\tDuration: 69.84 ms\tBilled Duration: 100 ms\tMemory Size: 1024 MB\tMax Memory Used: 106 MB\t\n",
  "parsedLambdaMessage": {
    "lines": [
      {
        "level": "info",
        "message": "/api/workspace/get/layer?locale=Mapbox&key=layer&layer=Mapbox%20Static&method=get",
        "timestamp": "2020-09-24T09:25:44.234Z"
      }
    ],
    "parse_status": "full",
    "report": {
      "billed_duration_ms": 100,
      "duration_ms": 70,
      "max_memory_used_mb": 106,
      "memory_size_mb": 1024
    },
    "request_id": "37854802-1647-4a7a-bc0b-0e15849e4831"
  },
  "path": "api/api",
  "projectId": "QmQ1CEwKy4SscKoWK1cfri7JPNHBpR1tW1RGYLoBtG6ExD",
  "proxy": {
    "clientIp": "86.162.182.233",
    "host": "xyz-mapbox.vercel.app",
    "method": "GET",
    "path": "/api/workspace/get/layer?locale=Mapbox&layer=Mapbox%20Static",
    "referer": "https://xyz-mapbox.vercel.app/",
    "region": "lhr1",
    "scheme": "https",
    "statusCode": 200,
    "timestamp": 1600939543895,
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36"
  },
  "requestId": "7ds59-1600939543895-c3f0d492de23",
  "source": "lambda",
  "statusCode": 200,
  "timestamp": 1600939544169
}
```

**Comment by chasers** [comment]
Commenting in case anyone else comes across this...

Sending straight to Logflare is useful in some instances but yes when you setup a log drain you can use `console`, `pino` or `pino-logflare` and Logflare attempts to parse those various payloads to give you something easily queryable. 

`pino-logflare` with the Vercel config example sets up `pino` such that client-side logs automatically get sent via HTTP and server-side logs get send to stdout which get picked up by Vercel's log drains.

I've put some examples up here: https://github.com/Logflare/next-pino-logflare-logging-example

**Reply by tdeitz** [reply]
@chasers so this explains why logged server side events (e.g. Nextjs API routes) down't show up in Logflare when running on localhost?
